### PR TITLE
Gallery polynomial fitting example

### PR DESCRIPTION
**Pull request by dkillick**
Adds a new gallery plot that demonstrates fitting a polynomial to a line plot of data from a cube.

### CODE REVIEW

File: docs/iris/example_code/General/polynomial_fit.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+"""
+Fitting a polynomial
+====================
+
+This example demonstrates adding a polynomial fit to a 1D plot of data from
```
Comment (ajdawson) [inline docs/iris/example_code/General/polynomial_fit.py]: I'd reword this because it sounds like the plot has to do with the fitting process (i.e. fitting to a plot). In fact the important part of this example is really the handling of metadata, so perhaps you could afford to be a little more verbose. Perhaps something along the lines of:

```
This example demonstrates computing a polynomial fit to 1D data from an Iris cube,
adding the fit to the cube's metadata, and plotting both the original data and the fit.
```

File: docs/iris/example_code/General/polynomial_fit.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+an Iris cube.
+
+"""
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+import iris
+import iris.plot as iplt
```
Comment (ajdawson) [inline docs/iris/example_code/General/polynomial_fit.py]: You don't need to import `iris.plot`.

File: docs/iris/example_code/General/polynomial_fit.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    location.add_aux_coord(fit, 0)
+
+    qplt.plot(location.coord('time'), location, label='data')
+    qplt.plot(location.coord('time'),
+              location.coord('polynomial_fit_of_data'),
+              'g-', label='polynomial fit')
+    plt.legend(loc='best')
+    plt.title('Trend of US air temperature over time')
+
+    iplt.show()
```
Comment (ajdawson) [inline docs/iris/example_code/General/polynomial_fit.py]: Use `qplt.show()` (or `plt.show()`), no need to import `iris.plot` just to do this.

File: docs/iris/example_code/General/polynomial_fit.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import iris
+import iris.plot as iplt
+import iris.quickplot as qplt
+
+
+def main():
+    fname = iris.sample_data_path('A1B_north_america.nc')
+    cube = iris.load_cube(fname)
+
+    # Extract a single time cross-section at a latitude and longitude point.
```
Comment (ajdawson) [inline docs/iris/example_code/General/polynomial_fit.py]: I wouldn't describe this as a cross section, that is usually reserved for a straight-line through a 2D space (i.e. has units of space rather than time). It should be referred to as a time-series.

File: docs/iris/example_code/General/polynomial_fit.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    location = next(cube.slices(['time']))
+
+    # Calculate a polynomial fit to the data at this cross-section.
```
Comment (ajdawson) [inline docs/iris/example_code/General/polynomial_fit.py]: As above for usage of cross-section.

File: docs/iris/example_code/General/polynomial_fit.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    x_points = location.coord('time').points
+    y_points = location.data
+    degree = 3
+
+    p = np.polyfit(x_points, y_points, degree)
+    f = np.poly1d(p)
+    y_fitted = f(x_points)
+
+    # Add the polynomial fit values to the cross-section to take
```
Comment (ajdawson) [inline docs/iris/example_code/General/polynomial_fit.py]: As above for usage of cross-section.

File: docs/iris/example_code/General/polynomial_fit.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    #  full advantage of Iris plotting functionality.
```
Comment (ajdawson) [inline docs/iris/example_code/General/polynomial_fit.py]: Remove the extra space at the beginning of this comment.

File: docs/iris/example_code/General/polynomial_fit.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    fname = iris.sample_data_path('A1B_north_america.nc')
+    cube = iris.load_cube(fname)
+
+    # Extract a single time cross-section at a latitude and longitude point.
+    location = next(cube.slices(['time']))
+
+    # Calculate a polynomial fit to the data at this cross-section.
+    x_points = location.coord('time').points
+    y_points = location.data
+    degree = 3
```
Comment (ajdawson) [inline docs/iris/example_code/General/polynomial_fit.py]: I'm not convinced this is a good choice for this example. A second degree polynomial fits this data pretty well too...

File: docs/iris/example_code/General/polynomial_fit.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+    p = np.polyfit(x_points, y_points, degree)
+    f = np.poly1d(p)
```
Comment (ajdawson) [inline docs/iris/example_code/General/polynomial_fit.py]: I wouldn't bother with constructing the polynomial then evaluating when you can do it in one step, given you only want to evaluate the fit with the one set of x values:

```python
y_fitted = np.polyval(p, x_points)
```

File: docs/iris/example_code/General/polynomial_fit.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    y_fitted = f(x_points)
+
+    # Add the polynomial fit values to the cross-section to take
+    #  full advantage of Iris plotting functionality.
+    fit = iris.coords.AuxCoord(y_fitted, long_name='polynomial_fit_of_data',
```
Comment (ajdawson) [inline docs/iris/example_code/General/polynomial_fit.py]: The `long_name` attribute should use the name of the quantity rather than just 'data': `polynomial_fit_of_air_temperature` or even `degree_3_polynomial_fit_of_air_temperature` (I don't mind either way, but if the second you could also consider including the degree using a format string and the variable `degree` defined previously).

File: docs/iris/example_code/General/polynomial_fit.py (hunk 3 above)
Comment (QuLogic) [inline docs/iris/example_code/General/polynomial_fit.py]: Don't use `plt.show()`; it breaks plotting in documentation.

File: docs/iris/example_code/General/polynomial_fit.py (hunk 1 above)
Comment (dkillick) [inline docs/iris/example_code/General/polynomial_fit.py]: Actually my primary intent with this example was to demonstrate the fitting process but I like the extension. Also the plot doesn't work anything like as well unless the fit is added to the cube's metadata!

File: docs/iris/example_code/General/polynomial_fit.py (hunk 4 above)
Comment (dkillick) [inline docs/iris/example_code/General/polynomial_fit.py]: Thanks :smile:

File: docs/iris/example_code/General/polynomial_fit.py (hunk 1 above)
Comment (ajdawson) [inline docs/iris/example_code/General/polynomial_fit.py]: >Also the plot doesn't work anything like as well unless the fit is added to the cube's metadata!

Well, it works just fine if you plot the fit AuxCoord against the time coord of the cube, you don't need to add the AuxCoord to the cube to achieve this (but I agree that you should in this case, to carry the new data in a context where it means something).

Comment (dkillick) [inline docs/iris/example_code/General/polynomial_fit.py]: > you don't need to add the AuxCoord to the cube to achieve this

Interesting! 

> but I agree that you should in this case, to carry the new data in a context where it means something

Quite.

File: docs/iris/example_code/General/polynomial_fit.py (hunk 3 above)
Comment (ajdawson) [inline docs/iris/example_code/General/polynomial_fit.py]: Why does it break using `plt.show()` and not `iplt.show()` or `qplt.show()`? Just curious, as these are just aliases.

File: docs/iris/example_code/General/polynomial_fit.py (hunk 9 above)
Comment (dkillick) [inline docs/iris/example_code/General/polynomial_fit.py]: You can indeed, but demonstrating both here has the benefit of introducing both functions to the readers of this example. I think `poly1d` might have a more beneficial and general application to the readers of this example.

File: docs/iris/example_code/General/polynomial_fit.py (hunk 10 above)
Comment (dkillick) [inline docs/iris/example_code/General/polynomial_fit.py]: Good point, I'll get changing that.

File: docs/iris/example_code/General/polynomial_fit.py (hunk 3 above)
Comment (dkillick) [inline docs/iris/example_code/General/polynomial_fit.py]: I think [this comment](https://github.com/SciTools/iris/pull/1356#issuecomment-58418007) explains it.

File: docs/iris/example_code/General/polynomial_fit.py (hunk 9 above)
Comment (ajdawson) [inline docs/iris/example_code/General/polynomial_fit.py]: I don't really agree. When users want to do more complex fitting where they need to re-use the polynomial and evaluate at multiple sets of points they should easily be able to find out how. I think adding an unnecessary step here is "hand-holding" and pushing users into learning about features of NumPy they *might* need later, but don't need right now to solve the problem at hand.

**Comment by dkillick** [comment]
Review actions (and a rebase).

File: docs/iris/example_code/General/polynomial_fit.py (hunk 1 above)
Comment (ajdawson) [inline docs/iris/example_code/General/polynomial_fit.py]: This still isn't quite right, you are implying the fit is to the plot, which it is not.

Comment (dkillick) [inline docs/iris/example_code/General/polynomial_fit.py]: It's amazing how much difference a word makes. I'm on it.

**Comment by dkillick** [comment]
And that's the missing bit from the docstring sorted.

**Comment by ajdawson** [comment]
Oh, I forgot... shouldn't this have a test?

**Comment by dkillick** [comment]
Indeed it should! And the latest commit has provided one :smile:

**Comment by ajdawson** [comment]
Looks good, can you squash these into a single commit @dkillick and I'll merge it.

**Comment by dkillick** [comment]
Commits squashed. Over to you @ajdawson!

**Comment by rhattersley** [comment]
Thanks chaps! :clap:

**Comment by ajdawson** [comment]
Good job @dkillick, thanks for doing this so quickly.

I'd been peer-reviewing a paper all day before I reviewed this PR so my critical judgement was turned up to maximum!

**Comment by dkillick** [comment]
No problem @ajdawson, thanks for the review and pointing out all those little bits I'd missed!

File: docs/iris/example_code/General/polynomial_fit.py (hunk 3 above)
Comment (QuLogic) [inline docs/iris/example_code/General/polynomial_fit.py]: Yep, that's exactly the reason. Obviously, there's no difference here because there's only one call, but for consistency's sake...
